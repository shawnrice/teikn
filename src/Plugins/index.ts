export * from './Plugin';
export * from './ColorTransformPlugin';
export * from './PrefixTypePlugin';
export * from './SCSSQuoteValuePlugin';
