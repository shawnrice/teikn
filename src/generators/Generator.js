const { EOL } = require('os');

class Generator {
  constructor(opts) {
    this.options = opts;
    this.validateOptions();
  }

  validateOptions() {
    const required = { ext: 'string' };

    const errors = [];

    Object.keys(required).forEach(name => {
      const type = typeof this.options[name];
      const expected = required[name];

      if (type !== expected) {
        errors.push(`Error: received option ${name} of type ${type}; expected ${expected}.`);
      }
    });

    if (errors.length) {
      throw new Error(errors.join(EOL));
    }
  }

  get file() {
    const { ext, filename = 'tokens' } = this.options;

    return [filename, ext].join('.');
  }

  header() {
    return null;
  }

  footer() {
    return null;
  }

  generateToken(token) {
    throw new Error('You need to extend the Generator class');
  }

  combinator() {
    throw new Error('You need to extend the Generator class');
  }

  generate(tokens) {
    return [this.header(), this.combinator(tokens), this.footer()]
      .filter(Boolean)
      .join(EOL)
      .trim();
  }
}

module.exports = Generator;
