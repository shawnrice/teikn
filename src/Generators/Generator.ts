import { EOL } from 'os';

import { version } from '../../package.json';
import { Plugin } from '../Plugins';
import { Token } from '../Token';
import { matches } from '../utils';

export type GeneratorOptions = {
  /**
   * The extension for the file
   *
   * (this is usually set when the Generator is extended)
   */
  ext: string;
  /**
   * The basename for the file
   *
   * default: `tokens`
   */
  filename?: string;
};

export type RequiredGeneratorOptions = {
  ext: string;
};

export type RequiredGeneratorOptionNames = keyof RequiredGeneratorOptions;

export abstract class Generator<Opts extends GeneratorOptions = GeneratorOptions> {
  options: Opts;

  constructor(opts: Opts) {
    this.options = opts;
    this.validateOptions();

    this.signature = this.signature.bind(this);
  }

  signature(): string {
    return `Teikn v${version}`;
  }

  validateOptions(): void {
    const required: RequiredGeneratorOptions = { ext: 'string' };

    const errors: string[] = [];

    for (const key in required) {
      const type = typeof this.options[key as RequiredGeneratorOptionNames];
      const expected = required[key as RequiredGeneratorOptionNames];

      if (type !== expected) {
        errors.push(`Error: received option ${key} of type ${type}; expected ${expected}.`);
      }
    }

    if (errors.length) {
      throw new Error(errors.join(EOL));
    }
  }

  get file(): string {
    const { ext, filename = 'tokens' } = this.options;

    return [filename, ext].join('.');
  }

  header(): string | null {
    return null;
  }

  footer(): string | null {
    return null;
  }

  protected prepareTokens(tokens: Token[], plugins: Plugin[]): Token[] {
    return tokens.map(token =>
      plugins.reduce((acc, plugin) => {
        if (!matches(plugin.tokenType, token.type)) {
          return acc;
        }

        if (!matches(plugin.outputType, this.options.ext)) {
          return acc;
        }

        return plugin.toJSON(acc);
      }, token),
    );
  }

  abstract generateToken(_: Token): any;

  abstract combinator(_: Token[]): string;

  generate(tokens: Token[], plugins: Plugin[] = []): string {
    return [this.header(), this.combinator(this.prepareTokens(tokens, plugins)), this.footer()]
      .filter(Boolean)
      .join(EOL)
      .trim();
  }
}

export default Generator;
