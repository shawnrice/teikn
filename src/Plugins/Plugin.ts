import { Token } from '../Generators/Token';

// export type HasTokenType = { tokenType: string | RegExp };

// export type HasOutputType = { outputType: string | RegExp };

// export type TransformFunction = (token: Token) => Token;

export abstract class Plugin {
  abstract tokenType: string | RegExp;
  abstract outputType: string | RegExp;

  options: { [key: string]: any };

  constructor(options: { [key: string]: any } = {}) {
    this.options = options;
  }

  abstract toJSON(token: Token): Token;
}
