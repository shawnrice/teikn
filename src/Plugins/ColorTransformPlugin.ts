import { Color } from '../Color';
import { Token } from '../Token';
import { Plugin } from './Plugin';

/**
 * Normalizes colors
 */
export class ColorTransformPlugin extends Plugin {
  outputType = /.*/;

  tokenType = 'color';

  constructor(options: { type?: 'rgb' | 'rgba' | 'hex' | 'hex3' | 'hsl' | 'hsla' | 'named' }) {
    super(options);
  }

  toJSON(token: Token) {
    const { type } = this.options;

    return {
      ...token,
      value: new Color(token.value).toString(type),
    };
  }
}
