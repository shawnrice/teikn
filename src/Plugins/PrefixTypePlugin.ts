import camelCase from 'lodash/camelCase';

import { Token } from '../Generators/Token';
import { Plugin } from './Plugin';

/**
 * Prefixes the token type to the token name
 *
 * E.g.
 *
 * ```javascript
 * {
 *   type: 'color',
 *   name: 'primary',
 *   value: 'orange'
 * }
 * ```
 * becomes
 * ```javascript
 * {
 *   type: 'color',
 *   name: 'colorPrimary',
 *   value: 'orange'
 * }
 * ```
 */
export class PrefixTypePlugin extends Plugin {
  outputType = /.*/;

  tokenType = /.*/;

  toJSON(token: Token) {
    const { type, name } = token;

    return {
      ...token,
      name: [camelCase(type), name[0].toUpperCase(), name.slice(1)].join(''),
    };
  }
}
