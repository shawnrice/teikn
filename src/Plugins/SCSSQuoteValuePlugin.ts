import { Token } from '../Token';
import { Plugin } from './Plugin';

/**
 * Prefixes the token type to the token name
 *
 * E.g.
 *
 * ```javascript
 * {
 *   type: 'color',
 *   name: 'primary',
 *   value: 'orange'
 * }
 * ```
 * becomes
 * ```javascript
 * {
 *   type: 'color',
 *   name: 'colorPrimary',
 *   value: 'orange'
 * }
 * ```
 */
export class SCSSQuoteValuePlugin extends Plugin {
  outputType = /s(a|c)ss/;

  tokenType = /^(font|font-family)$/;

  toJSON(token: Token) {
    const { value } = token;

    return {
      ...token,
      value: `unquote('#{${value}}')`,
    };
  }
}
