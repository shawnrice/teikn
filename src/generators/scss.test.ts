import { default as tokenSet1 } from '../fixtures/tokenSet1';
import { ColorTransformPlugin } from '../Plugins/ColorTransformPlugin';
import { PrefixTypePlugin } from '../Plugins/PrefixTypePlugin';
import { SCSSQuoteValuePlugin } from '../Plugins/SCSSQuoteValuePlugin';
import Generator from './SCSS';

describe('SCSSGenerator tests', () => {
  test('It has the correct filename', () => {
    expect(new Generator().file).toBe('tokens.scss');
  });

  test('It generates the token set', () => {
    expect(
      new Generator({ dateFn: () => 'null' }).generate(tokenSet1, [
        new ColorTransformPlugin({ type: 'rgba' }),
        new SCSSQuoteValuePlugin(),
        new PrefixTypePlugin(),
      ]),
    ).toMatchSnapshot();
  });
});
