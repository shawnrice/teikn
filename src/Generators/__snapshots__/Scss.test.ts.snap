// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SCSSGenerator tests It generates the token set 1`] = `
"///
/// Teikn v1.0.0-alpha.3
/// Generated null
///
/// This file is generated and should be commited to source control
///


// prettier-ignore
$token-values: (
  /// Use for body fonts
  font-family-body: unquote('#{\\"Roboto Condensed\\", Arial, sans}'),
  /// the primary branding color
  color-primary: rgba(240, 248, 255, 1),
);


/// Use \\"get-token\\" to access tokens by name
@function get-token($name) {
  @if (not map-has-key($token-values, $name)) {
    @error \\"Token '#{$name}' does not exist.\\";
  }
  @return map-get($token-values, $name);
}"
`;
